code/modules/mailorder/modr_catalogues.dm

```diff
@@ -0,0 +1,323 @@
+				if(length(src.canbuy) < 1)
+					. += "None!"
+				else
+					for(var/P in src.canbuy)
+						var/datum/mail_order/F = src.canbuy[P]
+						if(!istype(F, /datum/mail_order))
+							continue
+						var/itemct = length(F.order_items)
+						. += {"<a href='byond://?src=\ref[src];add_to_cart=[F.cleanname]'>[F.name]</a> - [itemct] Item(s) - $[F.cost]<br>
+						[F.desc]<br>"}
+						if(length(F.order_perm))
+							. += "Requires Access (1 or more of)"
+							for(var/acval in F.order_perm)
+								var/accessname = get_access_desc(acval)
+								. += " | [accessname]"
+							. += "<br>"
+						. += "<hr>"
+
+			if(MODE_CART)
+				. += "<h4>Shopping Cart</h4><br>"
+				if(length(src.cart) < 1)
+					. += "Empty - please use Catalogue."
+				else
+					var/entryct = length(src.cart)
+					. += "[entryct] Selections - [cartsize] Items - $[cartcost]<br>"
+					. += "<a href='byond://?src=\ref[src];checkout=1'>Check Out</a> | <a href='byond://?src=\ref[src];clearcart=1'>Clear Cart</a><br><hr>"
+					for(var/P in src.cart)
+						var/datum/mail_order/F = P
+						if(!istype(F, /datum/mail_order))
+							continue
+						var/itemct = length(F.order_items)
+						var/requiresid
+						if(length(F.order_perm))
+							requiresid = " | Requires ID"
+						else
+							requiresid = ""
+						. += {"[F.name]<br>
+						[itemct] Items | $[F.cost][requiresid]<br><hr>"}
+
+	Topic(href, href_list)
+		if(..())
+			return
+
+		if (href_list["viewlist"])
+			src.mode = MODE_LIST
+
+		if (href_list["viewcart"])
+			src.mode = MODE_CART
+
+		if (href_list["checkout"])
+			if(length(src.cart) > 0)
+				var/alert_beep = null
+				if(!src.master.host_program.message_silent)
+					alert_beep = src.master.host_program.message_tone
+
+				if (signal_loss >= 25)
+					src.master.display_alert(alert_beep)
+					var/displayMessage = "[bicon(master)] Unable to place order due to connection failure. Please try again later."
+					src.master.display_message(displayMessage)
+
+				else
+					var/creditCheck = src.authCard(usr)
+					if(creditCheck == "SUCCESS")
+						var/destination = "SHIP_TO_QM"
+						if(pick_landmark(LANDMARK_MAILORDER_SPAWN)) //pick a destination if mail insertion is supported by map
+							destination = input(usr, "Enter mail tag without quotes, or SHIP_TO_QM for secure crate-based delivery", src.name, null) as text
+						if(destination && isalive(usr))
+							var/final_bill = src.cartcost //da-na-na na, da-na-na na na
+							var/buy_success = src.shipCart(destination)
+							src.master.display_alert(alert_beep)
+							var/displayMessage = "[bicon(master)] Purchase unsuccessful due to lack of mail-order service to your area."
+							if(buy_success)
+								var/datum/db_record/spender = FindBankAccountByName(src.master.ID_card.registered)
+								if(spender)
+									spender["current_money"] -= final_bill
+								else
+									message_admins("<span class='alert'>[src] tried to charge a card that doesn't exist, yell at kubius</span>")
```

Comment:
This will also happen if it's an ID card with no associated name. For example if the HoP makes an ID with name "Smelly Stinkyface" for you. I don't want to get spammed as an admin if that happens.

Reply:
The code should stop before this point if there is no cardholder or no account on file, due to the authCard proc - this is present in case authCard malfunctions somehow.

Reply:
```suggestion
									CRASH("[src] tried to charge a card that doesn't exist, yell at kubius")
```